executor/executor.go

```diff
@@ -945,24 +973,42 @@ func (e *UnionExec) initialize(goCtx goctx.Context, forChunk bool) {
 			go e.fetchData(goCtx, i)
 		}
 	}
-	go e.waitAllFinished()
+	go e.waitAllFinished(forChunk)
 }
 
 func (e *UnionExec) resultPuller(childID int) {
 	defer e.wg.Done()
+	result := &unionWorkerResult{
+		err: nil,
+		chk: nil,
+		src: e.resourcePools[childID],
+	}
 	for {
 		if e.stopFetchData.Load().(bool) {
 			return
 		}
-		resource := <-e.resourcePool
-		resource.err = errors.Trace(e.children[childID].NextChunk(resource.chk))
-		if resource.err != nil || resource.chk.NumRows() == 0 {
-			if resource.err != nil {
-				e.stopFetchData.Store(true)
+		select {
+		case _, ok := <-e.finished:
+			if !ok {
+				return
+			}
+		case result.chk = <-e.resourcePools[childID]:
+		}
+		result.err = errors.Trace(e.children[childID].NextChunk(result.chk))
+		if result.err == nil && result.chk.NumRows() == 0 {
+			return
+		}
+		select {
+		case _, ok := <-e.finished:
+			if !ok {
+				return
 			}
+		case e.resultPool <- result:
```

Comment:
We can just send without `select`, the `resultPool` has enough capacity, it never blocks the sender.